### PR TITLE
SNS domain resolving

### PR DESCRIPTION
**Pull request by supermassive**
Resolves https://github.com/brave/brave-browser/issues/26652

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Run with `--enable-features=BraveWalletSns`
Omnibox navigation to to `onybose.sol` should redirect to `https://anirudha.co`

### CODE REVIEW

File: components/decentralized_dns/content/decentralized_dns_opt_in_page.cc
Comment (supermassive) [inline components/decentralized_dns/content/decentralized_dns_opt_in_page.cc:71]: Adjusted templates so they need only url params.
Removed `noopener noreferrer` flags as they are redundant on pages with opaque origin

**Review by iefremov — APPROVED** [review]
browser/net/* and chromium_src LGTM

**Comment by iefremov** [comment]
perhaps it may need a sec review

File: components/resources/decentralized_dns_strings.grdp
Comment (diracdeltas) [inline components/resources/decentralized_dns_strings.grdp:10]: why this change? we should not allow translators to inject HTML tags. actually i think there may already be a script that warns if there are any HTML tags in the translation files.

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (diracdeltas) [inline components/brave_wallet/browser/sns_resolver_task.cc:706]: why proceed with IPFS here? it seems like the user may not have opted into IPFS.

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (diracdeltas) [inline components/brave_wallet/browser/sns_resolver_task.cc:734]: is there a spec for these prefixes somewhere? like `IPFS.`

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc:706]: That comment actually about next line where we don't fail with error but set a flag and proceed with ipfs record fetching in `WorkOnDomainResolveTask`.

File: components/resources/decentralized_dns_strings.grdp
Comment (diracdeltas) [reply components/resources/decentralized_dns_strings.grdp:10]: ok, i verified `<a>` tags are allowed in `lib/l10n/validation.py`

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:726]: Internal error might be better, this is the result part, not the parameter.

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:760]: DITTO

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:766]: internal error maybe

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:785]: ditto

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:711]: Could you point to a reference to how url record looks like? And add link as comment here?

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (yrliou) [inline components/brave_wallet/browser/sns_resolver_task.cc:770]: A reference to how IPFS record looks like?

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc:706]: I think we should let user know that domain resolves to ipfs, but could not be navigated as it is disabled.
@yrliou wdyt?

Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc:706]: https://files.slack.com/files-pri/T04PX1BU8-F04AUF58FM3/image.png

Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc:706]: image

Comment (yrliou) [reply components/brave_wallet/browser/sns_resolver_task.cc:706]: I personally lean towards to not doing an additional check here, and just let the page fail to load if IPFS is disabled. I think it would be a more clear error as shown ^ to users.
And yes, it's correct that no requests to IPFS network will be issued until users opt in.

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc:771]: https://github.com/brave/brave-core/pull/15863/commits/cd6f31ee6d23c7c35177374065e5a69d9ddf6042

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc:711]: https://github.com/brave/brave-core/pull/15863/commits/cd6f31ee6d23c7c35177374065e5a69d9ddf6042

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15863/commits/cd6f31ee6d23c7c35177374065e5a69d9ddf6042

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15863/commits/cd6f31ee6d23c7c35177374065e5a69d9ddf6042

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15863/commits/cd6f31ee6d23c7c35177374065e5a69d9ddf6042

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15863/commits/cd6f31ee6d23c7c35177374065e5a69d9ddf6042

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/sns_resolver_task.cc]: https://github.com/brave/brave-core/pull/15863/commits/cd6f31ee6d23c7c35177374065e5a69d9ddf6042

File: components/brave_wallet/browser/json_rpc_service.cc
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_service.cc]: https://github.com/brave/brave-core/pull/15863/commits/cd6f31ee6d23c7c35177374065e5a69d9ddf6042

File: components/brave_wallet/browser/sns_resolver_task.cc
Comment (diracdeltas) [reply components/brave_wallet/browser/sns_resolver_task.cc:706]: sounds good to me

**Review by yrliou — APPROVED** [review]
code LGTM

**Comment by supermassive** [comment]
@diracdeltas can we merge this?
It seems there are no major security and privacy issues so we can have it in nightly behind a feature flag. We can address any further findings with 'enable feature by default' PR

**Comment by diracdeltas** [comment]
> It seems there are no major security and privacy issues so we can have it in nightly behind a feature flag. We can address any further findings with 'enable feature by default' PR

ok to go from security perspective but @ShivanKaul should take a look (he was busy yesterday). Is the plan to put SNS domain resolution ON by default? I thought it was 'Ask'?

**Comment by yrliou** [comment]
> > It seems there are no major security and privacy issues so we can have it in nightly behind a feature flag. We can address any further findings with 'enable feature by default' PR
> 
> ok to go from security perspective but @ShivanKaul should take a look (he was busy yesterday). Is the plan to put SNS domain resolution ON by default? I thought it was 'Ask'?

@diracdeltas I think what @supermassive meant was we will create another PR to actually enable the SNS feature flag when security & privacy review issue is closed, currently the feature flag is disable by default. And the resolve method is default to `Ask` and will not be changed. https://github.com/brave/brave-core/pull/15863/files#diff-3e5ca225627eb4399a80579331ce512deadf7355a7ec6ffb211bd7edda2de0fbR26-R27

**Comment by ShivanKaul** [comment]
If it's disabled behind feature flag I don't oppose merging now.

**Comment by diracdeltas** [comment]
ok sgtm to merge then and then open another PR with security review for enabling the feature flag by default

**Pull request merged**
